README.md
# decctools

decctools is an R package that provides easy access to energy statistics from the United Kingdom Department of Energy and Climate Change.  It currently provides access to two major data sources:
	
  * sub-national final energy consumption data at the local authority, middle super output area (MSOA), and lower super output area (LSOA) scales
	
  * fuel mix data and the associated carbon intensity of grid electricity using Elexon balancing mechanism data

You must have either [Curl](http://curl.haxx.se/) or [Wget](http://www.gnu.org/software/wget/) installed to use this package.

## Sub-national statistics

To access sub-national statistics, you will need to use one of three methods:

  * `get_LAD_data` gets the demands from Local Authority Districts.
	
  * `get_MSOA_data` gets the demands from Middle Super Output Areas
	
  * `get_LSOA_data` gets the demands from Lower Super Output Areas`
	
For all of these methods, see `?methodname` for options to select demands for only certain zones, sectors, or fuel types.  Note that not all geographies provide the same level of data; for example at the LSOA level, data are only available for domestic electricity and gas in England and Wales.

Here's an example:

    ## Gets energy data for electricity and gas use in the domestic sector in the most recent year
    df <- get_LAD_data(sector="domestic", fuel=c("electricity", "gas"))

## Grid carbon intensity

Elexon's [BM Reports](http://www.bmreports.com/bsp/bsp_home.htm) provides data on the UK's electricity market including the amount of electricity generated by different fuel types in each half-hourly period of the day.  Historic data (back to 2009-01-01) has been archived by the charity [Renewable Energy Foundation](http://www.ref.org.uk/fuel/) and decctools uses their summaries to provide the data used here.

There are two methods of interest:

  * `get_grid_mix(start, end)` gets the amount of electricity generated by different sources at half-hourly interval between two specified Dates
  
  * `get_grid_carbon(start, end)` gets the carbon intensity of grid electricity (kg CO2/kWh) between two specified Dates
  
An example:

    get_grid_carbon('2013-05-01', '2013-05-31')


NEWS
decctools 0.1.6
---------------

* Improved use of RCurl in response to CRAN comments

decctools 0.1.5
---------------

* Updated get_LAD_data URL

* Added error trapping for fetching of remote files

decctools 0.1.4
---------------

* Fixed get_last_date to cope with new REF website design

decctools 0.1.3
---------------

* Disabled curl's validation of SSL certificate to pass CRAN (Windows) checks.
  Not ideal but I can't reproduce the problem on my Windows machine. 

* Converted Depends statements to Imports

decctools 0.1.2
---------------

* updated URL for LAD data

decctools 0.1.1
---------------

* fixed bug with file download routines

decctools 0.1.0
---------------

* initial release

* provides methods for accessing UK Dept of Energy and Climate Change statistics on energy consumption at local authority district, MSOA, and LSOA level.

* provides methods for accessing grid fuel mix and carbon intensity of electricity using Elexon's BM Reports data, as archived by the charity Renewable Energy Foundation

R/lad.r
#' Get LAD energy consumption data
#'
#' Gets LAD (Local Administrative District) energy data from the DECC website, \url{http://www.decc.gov.uk}.
#' Since the format of these files tends to change just enough to be irritating for dependent code, there is
#' also an option that allows you to cache a local copy.
#'
#' @param id the unique id of the LAD to fetch.  If not specified, then all LADs are retrieved.
#' @param year the year to fetch. If not specified, then the most recent year is retrieved.  Only single years currently supported
#' @param sector a vector of economic sectors to fetch.  Valid values are 'domestic', 'industrial', 'total'.
#' @param fuel a vector of fuel types to fetch.  Valid values are 'coal', 'manufactured', 'petrol', 'gas', 'electricity', 'bioenergy', 'total'
#' @param dir an optional directory in which to cache the data
#' @return a long data frame with the requested data.  The 'energy' column is measured in GWh.
#' @keywords data energy
#' @export
#' @examples
#' \dontrun{
#' # Gets energy data for total fuels and sectors for most recent year
#' lad_data <- get_LAD_data()
#' }
#' 
#' \dontrun{
#' # Gets energy data for electricity and gas use in the domestic sector in the most recent year
#' # Depending on the status of DECC servers, this can sometimes fail.  In which case, an empty
#' # data frame is returned.
#' df <- get_LAD_data(sector="domestic", fuel=c("electricity", "gas"))
#' }
#'    
get_LAD_data <- function(id, year, sector='total', fuel='total', dir) {

  ## Download the data if necessary
  url <- "https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/274024/sub_national_total_final_energy_consumption_statistics_2005_2011.xlsx"
  file_name <- get_remote_file(url, dir)

  ## Load in the raw data from the spreadsheet
  wb <- tryCatch({
    loadWorkbook(file_name)
  }, error=function(e) {
    message(sprintf("Error loading workbook:\n\n%s\nTried download file from %s.  Email package maintainer to see if URL has changed.  Returning an empty data frame.", e, url))
    return(NULL)
  })

  ## If a valid workbook isn't found, return an empty data frame
  if (is.null(wb)) return(data.frame())
  
  ## Prep the year
  unit <- "GWh"
  valid_years <- get_valid_years(wb, unit)
  year_error <- FALSE

  ## Trap various error conditions
  if (missing(year)) {
    year_error <- TRUE
  } else if (is.na(year)) {
    year_error <- TRUE
  } else if (length(year)!=1) {
    warning("Only one year currently supported.  Defaulting to most recent year.")
    year_error <- TRUE
  } else if (!(year %in% valid_years)) {
    min_valid <- min(valid_years)
    max_valid <- max(valid_years)
    msg <- sprintf("%d is not a valid year.  Try a value in the range %d--%d.", year, min_valid, max_valid)
    warning(msg)
    year_error <- TRUE
  }
  
  if (year_error) year <- max(valid_years)
  
  ## Read in the data
  sheet_name <- paste(year, unit)
  df <- readWorksheet(wb, sheet_name, 4, 1, 421, 27)
  rm(wb)
  
  ## Since tmp contains all of the data, we now need to work some magic to get the right values
  df <- clean_decc_data(df, sector, fuel)

  ## Subset on target ids
  if (!missing(id)) {
    if (!is.na(id)) {
      df <- df[which(df$LAU1_code %in% id),]
    }
  }
  
  
  ## Renumber rows and return the result
  row.names(df) <- 1:nrow(df)
  return(df)
}


#' Determines which year values are valid
#' @param wb the workbook to search
#' @param unit the measurement unit of interest.  Defaults to GWh
#' @return a vector of years with data in the spreadsheet
#' @import stringr
get_valid_years <- function(wb, unit="GWh") {
  
  ## Get a list of the sheets
  sheets <- getSheets(wb)

  ## Keep only those with the units in the name
  sheets <- sheets[grep(unit, sheets)]

  ## Drop the unit suffix and make numeric
  years <- as.numeric(str_trim(str_replace(sheets, unit, "")))
  return(years)
}
  

#' Adds DECC energy consumption data to a specified data frame.  Currently
#' only selects domestic electricity and gas.
#'
#' @param df the raw DECC data frame from the spreadsheet
#' @param sector the sector to extract
#' @param fuel the fuels to extract
#' @return a long database with headers by sector and fuel
#' @import reshape2
clean_decc_data <- function(df, sector, fuel) {

  ## Manually define the column names
  sectors <- c(NA, NA, rep(c("industrial", "domestic", "total", NA), 2),
              "industrial", "domestic", "transport", "transport", "total", NA,
              rep(c("industrial", "domestic", "total", NA), 2),
              "total", NA, "total")
  fuels <- c(NA, NA, rep("coal", 3), NA, rep("manufactured", 3), NA,
             rep("petrol", 5), NA, rep("gas", 3), NA, rep("electricity", 3), NA,
             "bioenergy", NA, "total")


  ## Make sure we've got valid inputs
  sector <- unique(replace(sector, which(is.na(sector)), "total"))
  fuel <- unique(replace(fuel, which(is.na(fuel)), "total"))

  ## Calculate the columns that meet both criteria
  sector_cols <- which(sectors %in% sector)
  fuel_cols <- which(fuels %in% fuel)
  common_cols <- intersect(sector_cols, fuel_cols)
  
  ## Columns 1 and 2 are the LAU code and name
  good_cols <- c(1, 2, common_cols)  
  df <- df[, good_cols]

  ## Define the new column names
  sector_names <- sectors[common_cols]
  fuel_names <- fuels[common_cols]
  selection_cols <- paste(sector_names, fuel_names, sep=".")

  new_col_names <- c("LAU1_code", "name", selection_cols)
  names(df) <- new_col_names

  ## Now melt the data and final tidy
  df.m <- melt(df, id=c("LAU1_code", "name"), value.name="energy")
  df.m <- cbind(df.m, colsplit(df.m$variable, "\\.", c("sector", "fuel")))
  df.m <- df.m[,-3]
  df.m <- mutate(df.m, energy=as.numeric(str_replace(df.m$energy, "\\.\\.", NA)))
  ## Return the result
  return(df.m)
}

#' Gets the 2010 population estimates for all local authority districts
#'
#' This is a convenience function that returns the population data from \code{\link{LAD_metadata}}.
#'
#' @seealso \code{\link{LAD_metadata}} 
#' @export
#' @return a data frame with the LAU1_code and population
get_LAD_population <- function() {

  ## This is now stored in the embedded file so it's a quick job.
  return(LAD_metadata[,c("LAU1_code", "population")])

}
  


R/lsoa.r
#' Get LSOA energy consumption data
#'
#' This function fetches LSOA (Lower Super Output Area) data from the DECC website.
#'
#' @param id a vector of LSOA ids to fetch.  If not specified, then all LSOAs are retrieved.
#' @param fuel the fuel type to fetch.  Valid values are 'electricity', 'gas'
#' @param dir an optional directory in which to store a copy of the data
#' @return a long data frame with the requested data.  The 'energy' column is measured in GWh.
#' @keywords data energy
#' @export
#' @examples
#' \dontrun{
#' lsoa_data <- get_LSOA_data() # Gets all data
#' }
get_LSOA_data <- function(id, fuel=c("electricity", "gas"), dir) {

  ## As with the MSOA stuff, we'll prepare a list of all the parameters and then process them one by one.
  sector <- "domestic" # because at the moment, DECC doesn't provide other sectors
  
  ## Fetch the master parameters list
  params <- get_master_LSOA_params_list(dir)

  ## Subset to only those ones we care about
  cond <- lapply(params, function(l) return(l$sector %in% sector & l$fuel %in% fuel))
  params <- params[unlist(cond)]

  ## Now actually go and get the data
  tmp <- llply(params, parse_raw_LSOA_data)
  all_data <- do.call("rbind", tmp)

  ## Remove the unallocated LSOAs
  all_data <- all_data[all_data$LSOA_code!="Unallocated", ]

  ## Subset on the target ids
  if (!missing(id)) {
    if (!is.na(id))  {
      all_data <- all_data[which(all_data$LSOA_code %in% id),]
    }
  }

  ## Renumber rows and return the result
  row.names(all_data) <- 1:nrow(all_data)
  return(all_data)
    
}


## Now for each row in this data frame we need to get a data frame with the right data
##
## Params: 	l	a list giving the function parameters including
## url, ws, start_row, start_col, end_row, end_col, custom_function, data_col
parse_raw_LSOA_data <- function(l) {

  ## Get the file name (and download if necessary)
  file_name <- get_remote_file(l$url, l$dir)
  
  ## Load it into memory
  wb <- tryCatch({
      loadWorkbook(file_name)
  }, error=function(e) {
      message(sprintf("Error loading workbook:\n\n%s\nTried download file from %s.  Email package maintainer to see if URL has changed.  Returning an empty data frame.", e, l$url))
    return(NULL)
  })

  ## If a valid workbook isn't found, return an empty data frame
  if (is.null(wb)) return(data.frame())
  
  data <- readWorksheet(wb, l$sheet_name)
  
  ## Perform any custom changes to the data set
  lsoa <- data[,4]
  energy <- l$custom_function(data)
  
  ## Assemble the final data frame
  data <- data.frame(lsoa, energy, s=l$sector, f=l$fuel, row.names=1:length(lsoa))
  names(data) <- c("LSOA_code", "energy", "sector", "fuel")

  ## Convert kWh to GWh
  data <- mutate(data, energy=energy/1e6)
  
  ## Remove empty rows
  data <- data[!is.na(data$LSOA_code), ]

  return(data)
}

#' Builds a master set of parameters for MSOA data
#'
#' Creates a list of various parameters needed to download and extract
#' MSOA data from the DECC website.
#'
#' @param dir the directory in which to store a copy of the data
get_master_LSOA_params_list <- function(dir) {

  ## First specify the urls
  urls <- c("https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/49432/4813-llsoa-domestic-elec-est-2010.xls",
           "https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/49433/4814-llsoa-domestic-gas-est-2010.xls")

  ## And the corresponding worksheet names are
  worksheet_names <- c("LLSOA Electricity Domestic", "LLSOA Domestic gas")

  ## Define the sectors that these cover
  sectors <- rep("domestic", 2)
  fuels <- c("electricity", "gas")
  
  ## Then custom functions to retrieve the data since the domestic electricity is split
  ## These returns the code column first and then the energy value
  dom_gas_function <- function(df) return(df[,5])
  dom_elec_function <- function(df) return(apply(df[,5:6],1,sum,na.rm=TRUE))

  dir <- validate_directory(dir)
  
  ## Build a list summarizing everything
  df.l <- data.frame(url=urls, sheet_name=worksheet_names, sector=sectors, fuel=fuels,
                     dir=dir, stringsAsFactors=FALSE)
  df.l <- dlply(df.l, c("sheet_name"), as.list)
  
  ## Add the custom functions to to the list
  df.l <- lapply(df.l, function(l) c(l, list(custom_function=dom_gas_function)))
  df.l[[1]]['custom_function'] <- list(custom_function=dom_elec_function)
  return(df.l)

}

#' Gets the 2011 population estimates for all LSOAs
#'
#' This function gets the socio-demographic data associated with each Lower Super Output Area (LSOA). These data only cover England and Wales.
#'
#' @param dir an (optional) directory in which to save the downloaded data
#' @source \url{https://www.gov.uk/government/statistical-data-sets/socio-economic-data-for-mlsoa-igz-and-llsoa-electricity-and-gas-estimates}
#' @export
#' @return a data frame with the LSOA_code, population, area (in hectares), and number of households
get_LSOA_population <- function(dir) {

  ## Download the file
  url <- "https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/175644/Socio-economic_data_2013.xls"
  file_name <- get_remote_file(url, dir)

  ## Now open up the file and read the data
  ## Note that it is in two parts: English MLSOAs and Scottish IGZs
  wb <- tryCatch({
      loadWorkbook(file_name)
  }, error=function(e) {
      message(sprintf("Error loading workbook:\n\n%s\nTried download file from %s.  Email package maintainer to see if URL has changed.  Returning an empty data frame.", e, url))
      return(NULL)
  })

  ## If a valid workbook isn't found, return an empty data frame
  if (is.null(wb)) return(data.frame())
  
  pop_data <- readWorksheet(wb, "LLSOA England and Wales", startRow=2, startCol=3)
  rm(wb)

  ## Tidy up both files
  tidy_pop_data <- function(l) {
    names(l) <- c("LSOA_code", "name", "population", "area", "households")
    empty_rows <- which(is.na(l$population))
    if (length(empty_rows)>0) l <- l[-empty_rows,]
    return(l[,-2])
  }

  pop_data <- tidy_pop_data(pop_data)
  
  ## Return the result
  return(pop_data)
}
  


R/utils.r
#' Gets a file from a remote URL
#' 
#' Since the URLs of UK government data appear to change quite frequently, the method
#' will automatically cache a local copy.  
#'
#' @param url the address of the file to download
#' @param dir the directory to store the cached file in.  If undefined, it defaults to $TEMP/decctools.
#' @param update_cache boolean.  Should the cached file be overwritten?  Default = FALSE
#' @return the file name (path)
#' @import stringr RCurl
get_remote_file <- function(url, dir, update_cache=FALSE) {

  ## Currently only processes one URL
  if (length(url)>1) {
    warning("Only one URL at a time currently supported.  Using first URL only.")
    url <- url[1]
  } else if (is.na(url)) {
    stop("A valid URL must be specified.")
  } 

  dir <- validate_directory(dir)
  
  ## Split the URL to identify the filename
  tmp <- unlist(str_split(url, "/"))
  file_name <- file.path(dir, tmp[length(tmp)])

  ## If the file doesn't exist, then download it
  if (!file.exists(file_name) | update_cache) {

      ## see if the certificates are installed
      certs <- system.file("CurlSSL", "cacert.pem", package="RCurl")
      if (file.exists(certs)) {
          options(RCurlOptions=list(cainfo=certs))
      } else {
          warning(sprintf("Unable to find SSL certificates file.  Running without SSL validation.",
                          certs))
          options(RCurlOptions=list(ssl.verifypeer=FALSE))
      }

      if (!url.exists("http://www.google.com")) {
          stop("You must be connected to the internet to download this file.")
      } else {
          f <- CFILE(file_name, mode="wb")
          curlPerform(url=url, writedata=f@ref)
          close(f)
      }    
  }
  
  return(file_name)
}

#' Validates a user-specified directory
#'
#' Ensures that a user-specified directory exists.  If the argument is missing
#' then a placeholder directory is created in $TEMP.
#' @param dir the directory name
validate_directory <- function(dir) {
  ## Define the cache directory
  if (missing(dir)) dir=file.path(tempdir(), get_package_name())
  if (is.na(dir)) dir=file.path(tempdir(), get_package_name())

  ## Make sure it exists before trying to download
  if (!file.exists(dir)) dir.create(dir)

  return(dir)
}

#' Gets the name of this package
#'
#' @return a character string
get_package_name <- function() {
  return ("decctools")
}  

#' Gets a lookup table to match IDs across geographies
#'
#' Provides a lookup table giving the local authority name and corresponding IDs for LAU, LAD,
#' MSOA, and LSOA geographies.  Also returns a flag indicating whether or
#' not a given LAU is urban.  See \code{\link{LAD_metadata}} for more information.
#'
#' @param urban_classes classifications to be considered as urban.  Possible values include 'OU' (other urban), 'LU' (large urban), 'MU' (major urban), 'SR' (significant rural), 'R80' (80\% rural), 'R50' (50\% rural)
#' @param dir (optional) directory to save the lookup table information
#' @return a data frame
#' @export
#' @import stringr
get_lookup_table <- function(urban_classes=c("LU", "MU"), dir) {

  ## The lookup data is available in the LSOA electricity demands data from DECC
  ## so we'll use this rather than rooting around a master copy from ONS
  url <- "https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/49432/4813-llsoa-domestic-elec-est-2010.xls"

  file_name <- get_remote_file(url, dir)

  ## Load it into memory.  Only need first 4 columns
  wb <- tryCatch({
      loadWorkbook(file_name)
  }, error=function(e) {
      message(sprintf("Error loading workbook:\n\n%s\nTried download file from %s.  Email package maintainer to see if URL has changed.  Returning an empty data frame.", e, url))
      return(NULL)
  })

  ## If a valid workbook isn't found, return an empty data frame
  if (is.null(wb)) return(data.frame())
  
  data <- readWorksheet(wb, "LLSOA Electricity Domestic", startCol=1, endCol=4)
  names(data) <- c("LAU1_name", "LAU1_code", "MSOA_code", "LSOA_code")

  ## Get rid of empty lines
  data <- data[!is.na(data$LAU1_name), ]

  ## Also need to add in the Scottish IGZ codes
  ## Download the file
  url <- "https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/49424/4828-mlsoa-dom-elec-2010-alldata.xls"
  file_name <- get_remote_file(url, dir)

  ## Now open up the file and read the data
  ## Note that it is in two parts: English MLSOAs and Scottish IGZs
  wb <- loadWorkbook(file_name)
  scotland_data <- readWorksheet(wb, "MLSOA Domestic Electricity")
  rm(wb)
  scotland_data <- scotland_data[,1:3]
  names(scotland_data) <- c("LAU1_name", "LAU1_code", "MSOA_code")
  scotland_data <- scotland_data[!is.na(scotland_data$LAU1_name) &
                                 str_sub(scotland_data$MSOA_code, end=1)=="S", ]
  scotland_data <- cbind(scotland_data, LSOA_code=NA)

  ## Combine the two data sets
  data <- rbind(data, scotland_data)
  
  ## Some manual checking found a couple discrepancies in the names
  ## Tweak these to match
  ## unique(subset(tmp, is.na(LAU1_code) & !(country%in%c("Northern Ireland")))$LAU1_name)
  data <- mutate(data, LAU1_name=str_replace(data$LAU1_name, "Argyll \\& Bute UA Island", "Argyll and Bute"))
  data <- mutate(data, LAU1_name=str_replace(data$LAU1_name, "Edinburgh, City of", "City of Edinburgh"))
  data <- mutate(data, LAU1_name=str_replace(data$LAU1_name, "Dumfries \\& Galloway", "Dumfries and Galloway"))
  data <- mutate(data, LAU1_name=str_replace(data$LAU1_name, "Eilean Siar \\(Western Isles\\)", "Eilean Siar"))
  data <- mutate(data, LAU1_name=str_replace(data$LAU1_name, "Perth \\& Kinross", "Perth and Kinross"))
  data <- mutate(data, LAU1_name=str_replace(data$LAU1_name, "Cheshire West \\& Chester", "Cheshire West and Chester"))
  data <- mutate(data, LAU1_name=str_replace(data$LAU1_name, "Bristol, City of", "City of Bristol"))
  data <- mutate(data, LAU1_name=str_replace(data$LAU1_name, "Kingston upon Hull, City of", "City of Kingston upon Hull"))
  data <- mutate(data, LAU1_name=str_replace(data$LAU1_name, "Rhondda, Cynon, Taff", "Rhondda Cynon Taf"))
  data <- mutate(data, LAU1_name=str_replace(data$LAU1_name, "Vale of Glamorgan, The", "The Vale of Glamorgan"))

  ## Now load in the urban classification.  All the detailed processing for this is in the benchmarking paper
  ## This is lazy loaded as part of the package
  meta <- mutate(LAD_metadata, urban=(LAD_metadata$urban_class %in% urban_classes))
  meta <- meta[,c("name", "short_code", "long_code", "urban", "country")]

   ## Need to keep Scotland and Northern Ireland too
  tmp <- merge(data, meta, by.x="LAU1_name", by.y="name", all.y=TRUE) 
  return(tmp)
}


R/carbon-intensity.r
#' Gets mix of fuel sources in UK electricity
#'
#' Gets the mix of fuel sources used to generate UK grid electricity between specified dates.  Valid dates are yyyy-mm-dd and must be between 2009-01-01 and the current day; the start must come before the end date.
#'
#' @source The underlying data come from BM Reports, \url{http://www.bmreports.com/bsp/bsp_home.htm}, and are used under license from Elexon (see disclaimer below). However the BM Reports website is hard to use and so this function uses archived Elexon data compiled by the charity Renewable Energy Foundation and available at no cost at \url{http://www.ref.org.uk/fuel/}.
#' ELEXON does not and shall not be deemed to warrant or guarantee or make any representation (whether expressly or impliedly) as to the title, completeness, accuracy, reliability, freeness from error, method of preparation, merchantability or fitness for the purpose or for any particular purpose (whether or not any such purpose has been notified to ELEXON) of the Licensed Data and/or any use of or dealing with it.  ELEXON shall have no liability or responsibility whatsoever or howsoever arising (whether directly or indirectly) as a result of or in connection therewith, including in respect of each and all loss of profits, loss of revenue, loss of goodwill, loss of contracts, loss of business or production, indirect loss, economic or consequential loss, claims, demands, proceedings, actions, losses, liabilities, damages, costs, charges and expenses (whether or not any of the foregoing are foreseeable)
#'
#' @param start the start date for retrieving data.
#' @param end the end date for retrieving data
#' @return a data frame with a datetime stamp and average power generated by fuel type in MW
#' @examples
#' # These require a working internet connection
#' start <- "2010-01-01"
#' end <- "2010-01-03"
#' library(RCurl)
#' if (url.exists("http://www.google.com")) {
#'   data <- get_grid_mix(start, end) # Gets data for 1--3 January 2010
#' }
#' @export
#' @import stringr XML
get_grid_mix <- function(start, end) {

  ## Coerce strings to proper dates
  start <- as.Date(start)
  end <- as.Date(end)
    
  ## Validate inputs
  first_date <- as.Date("2009-01-01")
  last_date <- get_last_date()
  template <- "Invalid %s date.  Must be between %s and %s."
  if (start < as.Date("2009-01-01") | start > last_date) {
    stop(sprintf(template, "start", first_date, last_date))
  }

  if (end < as.Date("2009-01-01") | start > last_date) {
    stop(sprintf(template, "end", first_date, last_date))
  }

  if (end < start) stop("End date before start date.  Try again.")

  ## Create a sequence of dates
  dates <- seq(start, end, by="day")

  ## Load the data from the REF website
  tmp <- list(length(dates))
  columns <- c("character","numeric", rep("FormattedNumber", 14))
  base_url <- "http://www.ref.org.uk/fuel/tablebysp.php?valdate="
  for (i in 1:length(dates)) {
    url <- paste(base_url, dates[i], sep="")
    table <- suppressWarnings(readHTMLTable(url, colClasses=columns)[[2]])
    tmp[[i]] <- table
  }
  data <- do.call("rbind", tmp)

  ## Tidy things up
  names(data) <- str_trim(names(data))

  ## Convert date format into something more sensible
  data <- mutate(data, datetime=as.POSIXct((data$SP-1)*30*60, origin=data$SD))

  ## Rearrange
  data <- data[,c(17, 3:15)]
  return(data)
}


#' Calculates the carbon intensity of electricity generation
#'
#' Calculates the carbon intensity of electricity generation in the UK between specified dates.
#'
#' @seealso \code{\link{get_grid_mix}} and \code{\link{carbon_intensities}} for data sources
#' @param start the start date for retrieving data.
#' @param end the end date for retrieving data
#' @return a data frame with a datetime stamp and average carbon intensity of electricity measured in kg CO2/kWh
#' @examples
#' # These require a working internet connection
#' start <- "2010-01-01"
#' end <- "2010-01-03"
#' library(RCurl)
#' if (url.exists("http://www.google.com")) {
#'   carbon <- get_grid_carbon(start, end) # Gets grid carbon for 1--3 January 2010
#' }
#' @export
#' @import reshape2
get_grid_carbon <- function(start, end) {

  ## Get the grid mix data for this period
  data <- get_grid_mix(start, end)

  ## Merge with carbon intensities data for each fuel type
  data.m <- melt(data, id="datetime")
  data.m <- merge(data.m, carbon_intensities, by.x="variable", by.y="fuel")

  ## Calculate the total emissions in kg CO2
  ## cif = g CO2/kWh so value must be converted from MW in a half-hour period to kWh
  tmp <- mutate(data.m, emissions=(data.m$value*0.5)*1000*data.m$cif/1000)

  ## Now sum back up for a single time period
  tmp2 <- ddply(tmp, c("datetime"), summarize, demand=sum(get("value")*0.5*1000), total_em=sum(get("emissions")))
  result <- summarize(tmp2, datetime=tmp2$datetime, cif=tmp2$total_em/tmp2$demand)
  return(result)
}
  

#' Gets the most recent date for available data
#'
#' Gets the date of the last update to the REF fuels mix database.
#'
#' @return a Date object
#' @import lubridate
get_last_date <- function() {
  
  base_url <- "http://www.ref.org.uk/fuel/tablebysp.php"
  columns <- c("character","numeric", rep("FormattedNumber", 14)) 
  table <- suppressWarnings(readHTMLTable(base_url, colClasses=columns)[[2]])
  last_date <- as.Date(as.character(tail(table$SD, 1)))

  ## If the table isn't showing the data for some reason (which does happen),
  ## revert to the last day of the previous month
  if (length(last_date)==0) {
      tmp <- Sys.Date()
      day(tmp) <- 1
      day(tmp) <- day(tmp)-1
      last_date <- tmp
  }
      
  return(last_date)
}
                       
  


R/msoa.r
#' Get MSOA energy consumption data
#'
#' This function fetches MSOA (Middle Super Output Area) data from the DECC website from 2010.  Since the format of
#' these files tends to change just enough to be irritating for dependent code, there is also an option
#' that allows you to cache a local copy.
#'
#' @param id the unique id of the MSOA to fetch.  If not specified, then all MSOAs are retrieved.
#' @param sector the economic sector to fetch.  Valid values are 'domestic', 'nondomestic'
#' @param fuel the fuel type to fetch.  Valid values are 'electricity', 'gas'
#' @param dir an optional directory in which to save a copy of the data
#' @return a long data frame with the requested data.  The 'energy' column is measured in GWh.
#' @keywords data energy
#' @export
#' @examples
#' \dontrun{
#' msoa_data <- get_MSOA_data() # Gets all data
#' }
get_MSOA_data <- function(id, sector=c("domestic", "nondomestic"), fuel=c("electricity", "gas"), dir) {

  ## The easiest way to do this might be to just download all of the data, prepare the
  ## appropriate data frame and then return the subset.

  ## Because the format of each spreadsheet is slightly different we have to do some ugly hacking in
  ## the parse_raw_MSOA_data function below
  params <- get_master_MSOA_params_list(dir)
  
  ## Subset this to only those sectors that we care about
  cond <- lapply(params, function(l) return(l$sector %in% sector & l$fuel %in% fuel))
  params <- params[unlist(cond)]
  
  ## Now actually go and get the data
  tmp <- llply(params, parse_raw_MSOA_data)
  all_data <- do.call("rbind", tmp)

  ## Remove the unallocated MSOAs
  all_data <- all_data[all_data$MSOA_code!="Unallocated", ]

  ## Subset on the target ids
  if (!missing(id)) {
    if (!is.na(id)) {
      all_data <- all_data[which(all_data$MSOA_code %in% id),]
    }
  }
    
  ## Renumber rows and return the result
  row.names(all_data) <- 1:nrow(all_data)
  return(all_data)
}

## Now for each row in this data frame we need to get a data frame with the right data
##
## Params: 	l	a list giving the function parameters including
## url, ws, start_row, start_col, end_row, end_col, custom_function, data_col
parse_raw_MSOA_data <- function(l) {
  
  ## Get the file name (and download if necessary)
  file_name <- get_remote_file(l$url, l$dir)
    
  ## Load it into memory
  wb <- tryCatch({
      loadWorkbook(file_name)
  }, error=function(e) {
      message(sprintf("Error loading workbook:\n\n%s\nTried download file from %s.  Email package maintainer to see if URL has changed.  Returning an empty data frame.", e, l$url))
    return(NULL)
  })

  ## If a valid workbook isn't found, return an empty data frame
  if (is.null(wb)) return(data.frame())
                 
  data <- readWorksheet(wb, l$sheet_name)
  rm(wb)
  
  ## Perform any custom changes to the data set
  msoa <- data[,3]
  energy <- l$custom_function(data)

  ## Set the names
  data <- data.frame(msoa, energy, s=l$sector, f=l$fuel, row.names=1:length(msoa))
  names(data) <- c("MSOA_code", "energy", "sector", "fuel")

  ## Convert kWh to GWh
  data <- mutate(data, energy=energy/1e6)
  
  ## Remove empty rows and return
  data <- data[!is.na(data$MSOA_code), ]

  return(data)
}

#' Builds a master set of parameters for MSOA data
#'
#' Creates a list of various parameters needed to download and extract
#' MSOA data from the DECC website.
#'
#' @param dir the directory to use for saving the data
get_master_MSOA_params_list <- function(dir) {

    ## First specify the urls
  urls <- c("https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/49424/4828-mlsoa-dom-elec-2010-alldata.xls",
            "https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/49426/4842-mlsoa-dom-gas-2010-all-data.xls",
            "https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/49425/4830-mlsoa-non-dom-elec-2010-alldata.xls",
            "https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/49427/4831-mlsoa-non-dom-gas-2010-alldata.xls")

  ## And the corresponding worksheet names are
  worksheet_names <- c("MLSOA Domestic Electricity", "MLSOA Domestic Gas",
                      "MLSOA Non-domestic Electricity", "Non-domestic gas MLSOA")

  sectors <- c(rep("domestic", 2), rep("nondomestic", 2))
  fuels <- rep(c("electricity", "gas"), 2)
  
  ## Then custom functions to retrieve the data since the domestic electricity is split
  ## These returns the code column first and then the energy value
  null_function <- function(df) return(df[,4])
  dom_elec_function <- function(df) return(apply(df[,4:5],1,sum,na.rm=TRUE))

  dir <- validate_directory(dir)
  
  ## Build a list summarizing everything
  df.l <- data.frame(url=urls, sheet_name=worksheet_names, sector=sectors, fuel=fuels, dir=dir,
                     stringsAsFactors=FALSE)
  df.l <- dlply(df.l, c("sheet_name"), as.list)
  
  ## Add the custom functions to to the list
  df.l <- lapply(df.l, function(l) c(l, list(custom_function=null_function)))
  df.l[[1]]['custom_function'] <- list(custom_function=dom_elec_function)
  return(df.l)
}

#' Gets the 2011 population estimates for all MSOAs (including Scottish IGZs)
#'
#' This function gets the socio-demographic data associated with each Middle Super Output Area (MSOA).  
#'
#' @param dir an (optional) directory in which to save the downloaded data
#' @source \url{https://www.gov.uk/government/statistical-data-sets/socio-economic-data-for-mlsoa-igz-and-llsoa-electricity-and-gas-estimates}
#' @export
#' @return a data frame with the MSOA_code, population, area (in hectares), and number of households
get_MSOA_population <- function(dir) {

  ## Download the file
  url <- "https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/175644/Socio-economic_data_2013.xls"
  file_name <- get_remote_file(url, dir)

  ## Now open up the file and read the data
  ## Note that it is in two parts: English MLSOAs and Scottish IGZs
  wb <- tryCatch({
      loadWorkbook(file_name)
  }, error=function(e) {
      message(sprintf("Error loading workbook:\n\n%s\nTried download file from %s.  Email package maintainer to see if URL has changed.  Returning an empty data frame.", e, url))
      return(NULL)
  })

  ## If a valid workbook isn't found, return an empty data frame
  if (is.null(wb)) return(data.frame())
  
  england_data <- readWorksheet(wb, "MLSOA England and Wales", startRow=2, startCol=1)
  scotland_data <- readWorksheet(wb, "IGZ Scotland", startRow=2, startCol=1)
  pop_data <- list(england_data, scotland_data)
  rm(wb)

  ## Tidy up both files
  pop_data <- lapply(pop_data, function(l) {
    names(l) <- c("MSOA_code", "name", "population", "area", "households")
    empty_rows <- which(is.na(l$population))
    if (length(empty_rows)>0) l <- l[-empty_rows,]
    return(l[,-2])
  })
  pop_data <- do.call("rbind", pop_data)
  
  ## Return the result
  return(pop_data)
}
  
